### ISSUE TITLE
trouble generating ADSR envelopes from signal generator

### ISSUE THREAD

**Issue by jfayre** [naudio/NAudio#671]
Hello,
I need to generate various tones in my application. I'm trying to use the adsr envelope to generate a ping sound from a sine wave generated using the signal generator.
This is all working, but I'm wondering if the ADSR envelope sample provider is incomplete in the latest NAudio? I see properties for attack and release, but not sustain or decay. I can get the sine wave to fade in, but not start at full volume, then fade out.
I'm very new to c#, so would love a code sample if there is a better way to do this.
Thanks!

**Comment by markheath** [comment]
yes, the AdsrSampleProvider is incomplete - was just thrown in as a quick example of wrapping the envelope generator in a sample provider

**Comment by markheath** [comment]
Just following up on this - `AdsrSampleProvider` only exposed Attack and Release, with sustain pinned to full and decay to zero, so the "D" and "S" of the envelope were unreachable. That's now fixed in #1371, which surfaces the full envelope (`DecaySeconds` and `SustainLevel`, plus `Start()`, a `State` property, an `EnvelopeCompleted` event, and multi-channel support).

For the specific effect you were after — start at full volume, then fade out (a "ping") — you don't even need decay/sustain: use a tiny attack so it jumps to full level, let it hold, then call `Stop()` to trigger the release fade. The piece that was missing before was the `Stop()` call:

```csharp
var sine = new SignalGenerator(44100, 1)
{
    Frequency = 800, Gain = 0.5, Type = SignalGeneratorType.Sin
};

var adsr = new AdsrSampleProvider(sine)
{
    AttackSeconds  = 0.005f,   // near-instant rise -> starts at full volume
    ReleaseSeconds = 0.4f      // length of the fade-out
};

// play 'adsr' on your output device, then when you want the ping to end:
adsr.Stop();                   // enters Release and fades to silence
```

And with the new properties you can now build a full ADSR — e.g. a pluck that decays to a lower sustain level and then releases when the note ends:

```csharp
var adsr = new AdsrSampleProvider(sine)
{
    AttackSeconds = 0.01f,   // rise to full
    DecaySeconds  = 0.10f,   // fall to...
    SustainLevel  = 0.6f,    // ...this held level
    ReleaseSeconds = 0.40f   // fade-out after Stop()
};

// gate opens on construction by default; call adsr.Stop() for note-off.
```

If you're firing these as one-shots into a mixer, subscribe to `EnvelopeCompleted` (or check `State`) to know when a note has finished and can be removed.
